00XX-common-ledger-protocol/00XX-common-ledger-protocol.md

```diff
@@ -0,0 +1,273 @@
+
+- **CLP** is the common ledger protocol, as described by this packet and the
+  ASN.1 spec.
+
+- A **Side Protocol** is a protocol which isn't defined by CLP and is carried
+  in the side protocol data (see below).
+
+- A **CLP Connection** is a websocket connection over which CLP Packets are
+  sent.
+
+- **CLP Packets** are the messages defined in the CLP ASN.1 spec.
+
+- **Peers** are the parties on a CLP connection. Your peer is the party on the
+  other side of the CLP connection.
+
+- The **Bilateral Ledger** is the ledger which a CLP connection controls.  The
+  bilateral ledger is a persistent log of CLP packets, which can be used to
+deduce the current balance between two peers and the state of all transfers
+between them.
+
+- **Authoritative State** means that the entire Bilateral Ledger is being
+  tracked by a party. Both peers on a CLP connection can keep authoritative
+state, which means they can get into dispute if they disagree on the state of a
+transfer. This usually happens when network latency causes the peers to
+disagree about expiries. If one party keeps authoritative state, the other
+party must trust them not to tamper with it.
+
+## Overview
+
+CLP is broken up into 4 different RPC requests, which can get 2 different
+responses. Every CLP packet follows a common structure:
+
+```
+  +---------------+
+1 | Type (1)      |
+  +---------------+
+1 | Request ID    |
+2 | (2)           |
+3 |               |
+4 |               |
+  +---------------+
+1 | Length Prefix |
+2 | (3)           |
+  +---------------+
+  | Packet-       |
+  | Specific      |
+  | Data          |
+  . (4)           |
+  .               |
+  .               |
+  |               |
+  +---------------+
+1 | Side Protocol |
+2 | Count (5)     |
+  +---------------+
+  | Side Protocol |
+  | Data          |
+  . (6)           |
+  .               |
+  .               |
+  |               |
+  +---------------+
+```
+
+1. **Type**: A 1-byte value describing what type of CLP packet this is.
+The [CLP ASN.1 spec] describes what the different type IDs mean.
+
+2. **Request ID**: A 4-byte value used to correlate this requests and
+responses.
+
+3. **Length Prefix**: A 1-byte (if under 128) or 2-byte value, containing the
+combined length of the packet-specific data and side protocol data sections.
+
+4. **Packet-Specific Data:** Fields specific to the type of CLP packet. Variable
+length.
+
+5. **Side Protocol Count:** Variable-length integer containing the number of
+side protocols carried by this packet.
+
+5. **Side Protocol Data:** A list of key-value pairs, where the key is a
+length-prefixed ascii string, and the value is a length-prefixed octet string.
+The key signifies the side protocol's name, and the value is the data.
+
+## Flow
+
+CLP uses a simple RPC flow. A request-type CLP packet is sent, and a
+response-type CLP packet is sent in response with the same request ID.
+
+If a malformed CLP packet is sent, there should be no response. If a
```

Comment:
Might be worth mentioning why it's so important not to respond to unexpected responses

Reply:
Yeh, it seems odd to not return a response to a bad request.

Reply:
we should define what we mean with 'malformed CLP packet':
### unreadable packet (cannot answer):
* the first byte is outside the callset
* the total length is less than 5 bytes

### unexpected response (should not answer)
This is the one Evan refers to. The first byte indicates a response call (i.e., Ack, Response, or Error), and:
* the requestId is unknown to the recipient, or
* the requestId was used before in a well-formed request call, but after that the parser reached the end of the packet too early or too late.
* there is for instance an error code which is not listed in https://interledger.org/rfcs/0003-interledger-protocol/#errors

### unexpected request content (should answer with CLP-level error)
* the first byte indicates a request call (i.e., Message, Prepare, Fulfill, or Reject), and the next 4 bytes are a requestId that wasn't used before, but after that the parser reached the end of the packet too early or too late
* there is for instance an expiresAt that is in the past

Reply:
Agreeing with @michielbdejong. In particular, there should be a distinction between handling unexpected response-type packets and handling unexpected/malformed request-type packets.